File: music/src/main/java/com/xogrp/john/music/service/MusicPlayService.java
package com.xogrp.john.music.service;

import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.IBinder;
import android.os.RemoteException;
import android.util.Log;

import com.xogrp.john.music.activity.LockActivity;
import com.xogrp.john.toollibrary.utils.TextUtil;

import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.List;

/**
 * Created by Administrator on 2017/3/22.
 */

public class MusicPlayService extends Service {
    public static final String STOP_CMD = "com.xogrp.john.music.service.stop";
    public static final String NEXT_CMD = "com.xogrp.john.music.service.next";
    public static final String PLAY_CMD = "com.xogrp.john.music.service.play";
    public static final String PAUSE_CMD = "com.xogrp.john.music.service.pause";

    public static final String LOCK_SCREEN_CMD = "com.xogrp.john.music.service.lock";
    public static final String LOCK_SCREEN_CMD_FLAG = "com.xogrp.john.music.service.lock.flag";


    private static final String TAG = "ziq";
    private IBinder mBinder = new MusicPlayBinder(this);
    private int mCurrentSongPosition = -1;
    private List<MusicInfo> mMusicInfoList;
    private MediaPlayer mMediaPlayer = new MediaPlayer();
    private boolean isPause = false;
    private boolean isLockScreen = false;

    private MusicNotification mMusicNotification;
    @Override
    public void onCreate() {
        super.onCreate();
        Log.e(TAG, "onCreate: ");
        mMusicNotification = new MusicNotification(this);
        IntentFilter filter = new IntentFilter();
        filter.addAction(STOP_CMD);
        filter.addAction(NEXT_CMD);
        filter.addAction(PLAY_CMD);
        filter.addAction(PAUSE_CMD);
        filter.addAction(LOCK_SCREEN_CMD);
        filter.addAction(Intent.ACTION_SCREEN_OFF);
        registerReceiver(mIntentReceiver, filter);

    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.e(TAG, "onStartCommand: ");
        if (intent != null) {
            handleCommandIntent(intent);
        }
        return START_STICKY;
    }


    @Override
    public IBinder onBind(Intent intent) {
        Log.e(TAG, "onBind: ");
        return mBinder;
    }

    @Override
    public void onDestroy() {
        Log.e(TAG, "onDestroy: ");
        super.onDestroy();
    }

    private final BroadcastReceiver mIntentReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            handleCommandIntent(intent);
        }
    };

    public void handleCommandIntent(Intent intent){
        if(intent != null && !TextUtil.isTextEmptyOrNull(intent.getAction())){
            switch (intent.getAction()){
                case Intent.ACTION_SCREEN_OFF:
                    if(isPlaying() && !isLockScreen){
                        Intent lockActivity = new Intent(this, LockActivity.class);
                        lockActivity.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(lockActivity);
                    }
                    break;
                case STOP_CMD:
                    mMusicNotification.cancelNotification();
                    stop();
                    break;
                case NEXT_CMD:
                    next();
                    break;
                case PLAY_CMD:
                    play();
                    break;
                case PAUSE_CMD:
                    pause();
                    break;
                case LOCK_SCREEN_CMD:
                    isLockScreen = intent.getBooleanExtra(LOCK_SCREEN_CMD_FLAG, true);
                    break;
            }
        }
    }

    private static final class MusicPlayBinder extends MusicPlayServiceInterface.Stub{

        private WeakReference<MusicPlayService> musicPlayServiceWeakReference;

        public MusicPlayBinder(MusicPlayService service) {
            musicPlayServiceWeakReference = new WeakReference<MusicPlayService>(service);
        }

        @Override
        public void play() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                service.play();
            }
        }



        @Override
        public void next() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                service.next();
            }
        }

        @Override
        public void pause() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                service.pause();
            }
        }

        @Override
        public void stop() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                service.stop();
            }
        }

        @Override
        public boolean isPlaying() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                return service.isPlaying();
            }
            return false;
        }

        @Override
        public MusicInfo getCurrentMusicInfo() throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                return service.getCurrentMusicInfo();
            }
            return null;
        }

        @Override
        public void setMusicList(List<MusicInfo> musicList) throws RemoteException {
            MusicPlayService service = musicPlayServiceWeakReference.get();
            if(service != null){
                service.setMusicInfoList(musicList);
            }
        }
    }

    public MusicInfo getCurrentMusicInfo(){
        if(mMusicInfoList != null && mMusicInfoList.size() > 0 ){
            if(mCurrentSongPosition == -1){
                mCurrentSongPosition = 0;
            }
            return mMusicInfoList.get(mCurrentSongPosition);
        }
        return null;
    }

    public void play(){
        Log.e(TAG, "play: ");
        if(mMusicInfoList != null && mMusicInfoList.size() > 0 && !mMediaPlayer.isPlaying()){
            if(mCurrentSongPosition == -1){
                mCurrentSongPosition = 0;
            }
            if(isPause){
                mMediaPlayer.start();
                isPause = false;
            }else{
               playSong(mCurrentSongPosition);
            }
            mMusicNotification.updateNotification();
        }
    }

    public void next(){
        if(mMusicInfoList != null){
            int count = mMusicInfoList.size();
            if(count > 0 && mCurrentSongPosition == count - 1){
                mCurrentSongPosition = 0;
            }else{
                mCurrentSongPosition++;
            }
            playSong(mCurrentSongPosition);
        }
        mMusicNotification.updateNotification();
    }

    private void playSong(int position){
        if(mMediaPlayer != null){
            try {
                mMediaPlayer.reset();
//               通过ContentResolver 查出来的数据，是content：//  用Uri才能正确找到地址播放
//                Uri.parse(mMusicInfoList.get(position).data)
                mMediaPlayer.setDataSource(getApplicationContext(), Uri.parse(mMusicInfoList.get(position).data));
                mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
                // 通过异步的方式装载媒体资源
                mMediaPlayer.prepareAsync();
                mMediaPlayer.setOnPreparedListener(new MediaPlayer.OnPreparedListener() {
                    @Override
                    public void onPrepared(MediaPlayer mediaPlayer) {
                        mMediaPlayer.start();
                        mMusicNotification.updateNotification();
                    }
                });
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public void pause(){
        Log.e(TAG, "pause: ");
        if(mMediaPlayer != null){
            isPause = true;
            mMediaPlayer.pause();
        }
        mMusicNotification.updateNotification();
    }

    public void stop(){
        Log.e(TAG, "stop: ");
        if(mMediaPlayer != null){
            mMediaPlayer.stop();
        }
    }

    public boolean isPlaying(){
        return mMediaPlayer != null && mMediaPlayer.isPlaying();
    }

    public void setMusicInfoList(List<MusicInfo> musicInfoList) {
        this.mCurrentSongPosition = -1;
        this.mMusicInfoList = musicInfoList;
        if(mCurrentSongPosition == -1){
            mCurrentSongPosition = 0;
        }
    }


}

File: musicinkotlin/src/main/java/com/example/musicinkotlin/service/musicService/MusicPlayService.kt
package com.example.musicinkotlin.service.musicService

import android.app.Service
import android.content.*
import android.media.AudioManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Handler
import android.os.IBinder
import com.example.musicinkotlin.MusicPlayServiceBinder
import com.example.musicinkotlin.R
import java.lang.ref.WeakReference
import android.os.SystemClock





/**
 * Created by john on 06/12/2017.
 */
class MusicPlayService: Service() {

    private val mMediaPlayer: MediaPlayer = MediaPlayer()
    private var mMusicNotification: MusicNotification? = null

    private var mMusicList: MutableList<MusicInfo>? = mutableListOf()
    private var mCurrentMusicPos: Int = 0
    private var mIsPause: Boolean = false
    private val mHandler: Handler = Handler()

    companion object {
        fun startMusicPlayService(context: Context) {
            val intent = Intent(context, MusicPlayService::class.java)
            context.startService(intent)
        }

        fun bindMusicPlayService(context: Context, serviceConnection: ServiceConnection) {
            //need to start service first
            val intent = Intent(context, MusicPlayService::class.java)
            context.bindService(intent, serviceConnection, 0)
        }

        val PLAY_CMD = "com.example.musicinkotlin.service.musicService.play"
        val NEXT_CMD = "com.example.musicinkotlin.service.musicService.next"
        val PAUSE_CMD = "com.example.musicinkotlin.service.musicService.pause"
        val STOP_CMD = "com.example.musicinkotlin.service.musicService.stop"
        val REMOVE_NOTIFICATION_CMD = "com.example.musicinkotlin.service.musicService.removeNotification"
        const val UPDATE_STATUS_CMD = "com.example.musicinkotlin.service.musicService.updateStatus"
    }

    override fun onCreate() {
        super.onCreate()
        mMusicNotification = MusicNotification(this)
        val intentFilter: IntentFilter = IntentFilter()
        intentFilter.addAction(PLAY_CMD)
        intentFilter.addAction(NEXT_CMD)
        intentFilter.addAction(PAUSE_CMD)
        intentFilter.addAction(STOP_CMD)
        intentFilter.addAction(REMOVE_NOTIFICATION_CMD)
        registerReceiver(mBroadcastReceiver, intentFilter)
    }

    //匿名内部类
    private var mBroadcastReceiver: BroadcastReceiver? = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            handleIntent(intent)
        }
    }

    fun handleIntent(intent: Intent?): Unit {
        val action = intent?.action
        when(action){
            PLAY_CMD -> play()
            NEXT_CMD -> next()
            PAUSE_CMD -> pause()
            STOP_CMD -> stop()
            REMOVE_NOTIFICATION_CMD ->{
                mMediaPlayer.stop()
                mMusicNotification?.removeNotification()
            }
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        handleIntent(intent)
        return START_NOT_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? {
        return MusicBinder(this)
    }

    override fun onDestroy() {
        unregisterReceiver(mBroadcastReceiver)
        super.onDestroy()
    }

    fun updateNotification(): Unit {
        mMusicNotification?.updateNotification()
        sendBroadcast(Intent(UPDATE_STATUS_CMD))
    }

    private val mTicker = object : Runnable {
        override fun run() {
            sendBroadcast(Intent(UPDATE_STATUS_CMD))
            val now = SystemClock.uptimeMillis()
            val next = now + (1000 - now % 1000)
            mHandler.postAtTime(this, next)
        }
    }


    fun play(): Unit {
        if(!mMediaPlayer.isPlaying){
            if(mIsPause){
                mIsPause = false
                mMediaPlayer.start()
                updateNotification()
            }else{
                playSong()
            }
        }
    }

    private fun playSong(): Unit {
        val musicUrl = getCurrentMusicInfo()?.uri
        mMediaPlayer.reset()
        mMediaPlayer.setDataSource(applicationContext, musicUrl)
        mMediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC)
        // 通过异步的方式装载媒体资源
        mMediaPlayer.prepareAsync()
        mMediaPlayer.setOnPreparedListener {
            mMediaPlayer.start()
            mHandler.post(mTicker)
            updateNotification()
        }
        mMediaPlayer.setOnCompletionListener(MediaPlayer.OnCompletionListener {
            next()
        })
    }

    fun next(): Unit {
        val count = mMusicList?.size!!
        mCurrentMusicPos++
        if(mCurrentMusicPos >= count){
            mCurrentMusicPos = 0
        }
        playSong()
    }

    fun getCurrentPosition(): Int {
        return 100 * mMediaPlayer.currentPosition / mMediaPlayer.duration
    }

    fun pause(): Unit {
        if (mMediaPlayer.isPlaying){
            mIsPause = true
            mMediaPlayer.pause()
            updateNotification()
        }
    }

    fun stop(): Unit {
        mMediaPlayer.stop()
        updateNotification()
    }

    fun isPlaying(): Boolean {
        return mMediaPlayer.isPlaying
    }

    fun setMusicList(musicList: MutableList<MusicInfo>?) {
        mMusicList = musicList
    }

    fun getCurrentMusicInfo(): MusicInfo? {
        return mMusicList?.get(mCurrentMusicPos)
    }


    class MusicBinder(): MusicPlayServiceBinder.Stub(){
        private var mService: WeakReference<MusicPlayService>? = null

        constructor(service: MusicPlayService) :this(){
            mService = WeakReference(service)
        }

        override fun play() {
            mService?.get()?.play()
        }

        override fun next() {
            mService?.get()?.next()
        }

        override fun getCurrentPosition(): Int {
            return mService?.get()?.getCurrentPosition()!!
        }

        override fun pause() {
            mService?.get()?.pause()
        }

        override fun stop() {
            mService?.get()?.stop()
        }

        override fun isPlaying(): Boolean {
            return mService?.get()?.isPlaying()!!
        }

        override fun getCurrentMusicInfo(): MusicInfo? {
            return mService?.get()?.getCurrentMusicInfo()
        }

        override fun setMusicList(musicList: MutableList<MusicInfo>?) {
            mService?.get()?.setMusicList(musicList)
        }
    }



}